llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp

```diff
@@ -1470,12 +1470,11 @@ Value *SCEVExpander::expand(const SCEV *S) {
 
   // We can move insertion point only if there is no div or rem operations
   // otherwise we are risky to move it over the check for zero denominator.
-  auto SafeToHoist = [](const SCEV *S) {
-    return !SCEVExprContains(S, [](const SCEV *S) {
+  auto SafeToHoist = [&](const SCEV *S) {
+    return !SCEVExprContains(S, [&](const SCEV *S) {
               if (const auto *D = dyn_cast<SCEVUDivExpr>(S)) {
-                if (const auto *SC = dyn_cast<SCEVConstant>(D->getRHS()))
-                  // Division by non-zero constants can be hoisted.
-                  return SC->getValue()->isZero();
+                if (SE.isKnownNonZero(D->getRHS()))
```

Comment:
Do you also need to prove it's not poison?

Reply:
That is a great question.  I think we do.  We don't in a whole bunch of other cases, so this bug is not unique my new code, but yeah, I don't see any reason why hosting "udiv X, poison" above control flow would be legal.

Reply:
Most places that check value ranges don't care whether the input is poison: if the input is poison, the output is poison anyway.  I guess there might be some mistakes elsewhere, though.

Reply:
Every place we consider hoisting a udiv appears to be wrong by this logic.  We also appear to need to worry about undef by the same logic.  We have several different instances in SCEV Expander alone.  

We also have no infrastructure for proving a SCEV non-poison directly, so the scope of a fix here is significant.  I think we'd basically have to duplicate all of isGuaranteedNotToBeUndefOrPoison over SCEV.

I'm not going to drive this forward as this patch was merely a small fix observed while exploring something else, and the effort doesn't appear justified

Reply:
>  We also appear to need to worry about undef by the same logic.

Undef is not known to be non-zero.

> We also have no infrastructure for proving a SCEV non-poison directly, so the scope of a fix here is significant. I think we'd basically have to duplicate all of isGuaranteedNotToBeUndefOrPoison over SCEV.

We have the infrastructure for this (SCEVPoisonCollector), but it looks like we currently don't expose an isGuaranteedNotToBePoison() API based on it.